### ISSUE TITLE
In DITA-OT project files, allow subsets of deliverables to be published

### ISSUE THREAD

**Issue by chrispy-snps** [dita-ot/dita-ot#3914]
## Description
Currently, DITA-OT project file publishing is limited to publishing one deliverable or all deliverables.

To publish subsets of deliverables, the `<deliverable>` definitions must be split into multiple files, with subsets built using a hierarchy of `<include>` directives:

![files](https://user-images.githubusercontent.com/50950969/163050643-a1296063-e3c8-4b04-98ca-5e140ba0e2f6.svg)

This has the following drawbacks:

* Many files can be required.
  * We have 100+ contexts, shared by dozens of products, organized into several help collections.
  * Our deliverables are resistant to grouping in fewer files, due to being shared where products overlap.
* It is difficult to know what a particular file publishes without tracing through multiple files.
* It is difficult to explore the evolution of project file content over time in a revision control system.
* Maintenance operations (search-and-replace, adjusting collections) are more complex.
* `<context>` and `<publication>` information must be replicated in every deliverable file to allow individual publishing (inelegant, but technically harmless).

## Possible Solution
It would be useful to be able to natively define collections of deliverables (and collections of collections), perhaps like this:

![collections](https://user-images.githubusercontent.com/50950969/163050681-e67b991d-eda4-4181-9d5b-1aa528e735b5.svg)

where a "collection" represents its subset of referenced deliverables:

```
dita --project project.xml --deliverable product3
```

When a collection is published, its deliverables should be published in the order resulting from expanding the references. This provides support for order-dependent deliverables (such as HTML online help that cleans its output directory first, followed by PDF deliverables written to that same output directory).

If the reference expansion includes a deliverable multiple times, the deliverable should only be published once (the first occurrence seems reasonable).

Detection for circular collection reference loops would be needed.

## Potential Alternatives
I considered having a deliverable declare its dependencies via a new `<depends>` element, but this made it difficult to specify whether to publish a deliverable alone or with its dependencies, plus there was no intuitive way to describe order dependency.

## Additional Context
A testcase is included:

[ditaot_project_file_collections.zip](https://github.com/dita-ot/dita-ot/files/8476477/ditaot_project_file_collections.zip)

The `project.xml` file in the testcase uses the format proposed above, although other implementations are possible.

**Comment by raducoravu** [comment]
This also sounds useful to me.

**Comment by chrispy-snps** [comment]
@xephon2 - do I recall correctly that this was something interesting to you? If so, what are your thoughts on this enhancement request?

**Comment by xephon2** [comment]
@chrispy-snps , not much to add. This is exactly how it should work.

**Comment by jelovirt** [comment]
In project files, you have `context`, `publication`, and `output` and those make up a `deliverable`. Multiple `deliverable` can be grouped into a `collection`. Yeah.

But is the keyword `collection` right for a set of deliverables. My I would expect it to be a _collection of contexts_. Maybe because in MongoDB—which I use at work—there’s are collections that contain documents.

`deliverable-set` would be one, but I don’t like it and would like to avoid hyphenated terms.

Regardless of the keyword chosen, the feature seems simple enough: a named set of deliverable references.

**Comment by jelovirt** [comment]
Syntax equivalence:

# XML

```xml
<project>
  <collection id="product1" name="Product 1">
    <deliverable idref="A"/>
    <deliverable idref="B"/>
  </collection>
</project>
```

# YAML

```yaml
collections:
- name: Product 1
  id: product1
  deliverables:
  - idref: A
  - idref: B
```

# JSON
```json
{
  "collections": [{
    "name": "Product 1",
    "id": "product1",
    "deliverables": [
      { "idref": "A" },
      { "idref": "B" }
    ]
  }]
}
```

**Comment by xephon2** [comment]
**Excursus**:

I generated an iiRDS package with two dummy PDFs using the [iiRDS Open Toolkit](https://iirds-ot.plusmeta.de/iirds-open-toolkit/generate-iirds). The iiRDS OT is a public webservice, just click'n'play if you like. The iiRDS content delivery package design is slightly similar to this _DITA-OT Deliverable Collection_ approach. An iiRDS package groups assets and defines their semantical relationships with RDF.

In this syntax it works the other way round. Instead of defining a non-semantical collection which simply groups the assets, the asset "_knows_" the semantical relations to their peers and parents.

**Why could it matter?**

Jarno is, for good reasons, always very careful when it comes to backwards compatibility. This means, we should think about the long-term vision of the DITA-OT project file concept. One way could be to think of it more as a replacement for a build script (for example an Ant file). Another way would be to give it a broader, high-level role, as a semantical container for DITA publishing maps, so, a higher level of abstraction to define semantical relationships across deliverables. 

Personally I think that the definition of relations between deliverables should be more a topic for the OASIS DITA TC (maybe for 3.0) and not for the DITA-OT. For the DITA-OT I think the DITA-OT project file is and should be seen as a replacement for a build script and should not contain semantical data.

```xml
<?xml version="1.0"?>
<rdf:RDF xmlns:iirds="http://iirds.tekom.de/iirds#"
  xmlns:iirdsMch="http://iirds.tekom.de/iirds/domain/machinery#"
  xmlns:iirdsSft="http://iirds.tekom.de/iirds/domain/software#"
  xmlns:plus="https://data.plusmeta.de/metadata#"
  xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
  xmlns:rdfs="http://www.w3.org/2000/01/rdf-schema#"
  xmlns:xml="http://www.w3.org/XML/1998/namespace">

  <!-- *** iiRDS Open Toolkit (v1.1.1) *** -->
  <!-- *** generated on 14.4.2022, 09:27:50 *** -->
  <iirds:Package rdf:about="urn:uuid:824ea311-6495-4d13-9c77-b481af9ee844/package">
    <iirds:title>iiRDS Open Toolkit Export</iirds:title>
    <iirds:iiRDSVersion>1.1</iirds:iiRDSVersion>
  </iirds:Package>

  <iirds:Document rdf:about="urn:uuid:34a74b6f-2211-47ea-8bf3-51d3d76c6712">
    <iirds:is-part-of-package rdf:resource="urn:uuid:824ea311-6495-4d13-9c77-b481af9ee844/package"/>
    <iirds:has-rendition>
      <iirds:Rendition rdf:about="urn:uuid:34a74b6f-2211-47ea-8bf3-51d3d76c6712/rendition/foo.pdf">
        <iirds:format>application/pdf</iirds:format>
        <iirds:source>CONTENT/foo.pdf</iirds:source>
      </iirds:Rendition>
    </iirds:has-rendition>
  </iirds:Document>

  <iirds:Document rdf:about="urn:uuid:ce622607-c59b-46a3-bf8f-901f2337ada1">
    <iirds:is-part-of-package rdf:resource="urn:uuid:824ea311-6495-4d13-9c77-b481af9ee844/package"/>
    <iirds:has-rendition>
      <iirds:Rendition rdf:about="urn:uuid:ce622607-c59b-46a3-bf8f-901f2337ada1/rendition/bar.pdf">
        <iirds:format>application/pdf</iirds:format>
        <iirds:source>CONTENT/bar.pdf</iirds:source>
      </iirds:Rendition>
    </iirds:has-rendition>
  </iirds:Document>

</rdf:RDF>
```

**Comment by xephon2** [comment]
Regarding the element name, @jelovirt, I'd suggest to use `<package>` to avoid `<collection>` and `<deliverable-set>`. This would also be consistent with the iiRDS terminology.

**Comment by chrispy-snps** [comment]
I think different terms have different connotations to different people. For me, `<package>` has a software/OO connotation that does not intuitively apply to deliverables, whereas our team uses "help collection" to refer to HTML+PDF outputs, so that is intuitive to me.

`<group>` is another option. But no matter what fourth word we choose, its relationship to `<context>`, `<publication>`, and `<deliverable>` might not be obvious to some people.

I tend to avoid hyphenated elements too. However, DITA uses them (`<draft-comment>`, `<related-links>`, `<required-cleanup>`, etc.). And in this case, `<deliverable-set>` would declare its purpose and relationship. The more I think about it, the more it overcomes my reluctance.

Regarding parent/peer semantics, originally I thought a deliverable could define its dependencies:

```
<deliverable id="olh">
  <context .../>
  <output .../>
  <publication .../>

  <requires idref="A"/>
  <requires idref="B"/>
  <requires idref="C"/>
</deliverable>
```

But I encountered problems with this:

* How would you specify whether to publish "olh" alone or with its dependencies?
* How would you define peer groups of deliverables with no clear parent?
* How do you specify ordering - should A+B+C be built first or last?

This led me to the idea of "empty" deliverables:

```
<deliverable id="olh-full">

  <!-- no context/output/publication -->

  <requires idref="olh"/>
  <requires idref="A"/>
  <requires idref="B"/>
  <requires idref="C"/>
</deliverable>
```

which then turned into this enhancement idea.

I also view project files as a standardized build script. I haven't figured out how to do preprocessing/postprocessing yet. But with collections, preprocessing/postprocessing could be implemented via a plugin and a "contentless" deliverable, sequenced in with the required ordering:

```
<publication id="compress-pdfs" transtype="compress-pdfs"/>

<deliverable id="olh-full-compress-pdfs">
  <context ???/>  <!-- not sure, there really isn't one -->
  <output href="olh-dir"/>
  <publication idref="compress-pdfs"/>
</deliverable>

<deliverable-set id="olh-full">
  <deliverable idref="olh"/>
  <deliverable idref="A"/>
  <deliverable idref="B"/>
  <deliverable idref="C"/>
  <deliverable idref="olh-full-compress-pdfs"/>
</deliverable-set>
```

A full exploration probably belongs somewhere else. But if the future could allow _actions_ to be sequenced with deliverables (truly like a build script), then perhaps this could affect our choice of "collection" element name now?

I share the concerns about backward compatibility. Hopefully we can thoughtfully identify all the new functionality that is needed. And all things considered, the "less-backwards-compatible" version jump to DITA-OT 4.0 provides a good opportunity to do it!

**Comment by chrispy-snps** [comment]
A DITA authoring/CCMS tool could present collections-of-collections, collections, and deliverables graphically in expandable tree form, making it easy to explore and select from the available publishing options.

**Comment by jelovirt** [comment]
Thinking about the syntax and trying to make it more compact.

Syntax equivalence:

# XML

```xml
<project>
  <collection id="product1" name="Product 1">
    <deliverable idref="A"/>
    <deliverable idref="B"/>
    <collection idref="product2"/>
    <collection idref="product3"/>
  </collection>
</project>
```

This could be more compact by using attributes with comma separated list

```xml
<project>
  <collection id="product1" name="Product 1"
              deliverables="A, B"
              collections="product2, product3"/>
</project>
```

But I don't really like this.

# YAML

```yaml
collections:
- name: Product 1
  id: product1
  deliverables:
  - A
  - B
  collections:
  - product2
  - product3
```

# JSON
```json
{
  "collections": [{
    "name": "Product 1",
    "id": "product1",
    "deliverables": [ "A", "B"],
    "collections": [ "product2", "product3" ]
  }]
}
```

**Comment by jelovirt** [comment]
Is the order of deliverable generation a requirement (👍🏽/👎🏽)?

If we call this a _collection set_, then it's unordered; if the processing order needs to be configurable, then using the word _set_ might cause confusion.

**Comment by chrispy-snps** [comment]
@jelovirt - syntax-wise, I prefer the `@id`/`@idref` approach for the following reasons:

- It is consistent with how project file references work today.
- XSLT preprocessing is easier because of the built-in `id()` and `idref()` functions.
- In revision control history, the addition or removal of references are line-based changes that more easily merge, revert, etc.

**Comment by chrispy-snps** [comment]
@jelovirt - I am normally not a fan of hyphens in element names, but I have come to prefer `<deliverable-set>` because

- It makes more sense with `dita --deliverable`.
- It keeps terminology focused on the "big three" - contexts, publications, deliverables - without introducing a fourth term.
- Its purpose and relationship to the "big three" are more intuitively clear without referring back to documentation.

So regardless of the ordering connotation that comes with `<deliverable-set>`, I have come to prefer it.

An ordering guarantee for deliverable generation is a nice-to-have, but not a requirement.

For us, the documentation for a product family is (1) one Oxygen WebHelp deliverable that contains all the books as submaps, plus (2) individual book PDFs placed into that same output directory:

![html5_and_pdf](https://user-images.githubusercontent.com/50950969/171863038-3900508f-d3fe-491c-8b52-311b1e1e1081.svg)

Because WebHelp is HTML-based, we must clean the output directory before publishing it to avoid orphaned files (see #1199), but that must be done before the PDF deliverables are also written into that directory.

If there were an ordering guarantee "built into" deliverable sets, then we could invoke output-cleaning plugins in a certain way. If not, then output-cleaning must be moved to a preprocessing operation that computes and cleans all output directories for the deliverables to be published. This is easy in automated linux publishing where I can write a wrapper script to do that, but not easy in Oxygen where writers interactively publish from the software's UI. They will need to remember to manually delete their output directory from time to time.

And some day, perhaps if the DITA-OT implements a "parallel deliverable publishing" capability (wouldn't that be cool!), then an ordering guarantee becomes even harder to implement because you'd need to derive a whole `make`-like dependency tree, and that is asking a lot for a feature you are trying to keep simple!

So if an ordering guarantee is implemented, I think value could be obtained from it. But it's just a nice-to-have, not a requirement.

@xephon2 - do you have any thoughts on order of deliverable generation in a deliverable set?
